### PR TITLE
Transposition

### PR DESCRIPTION
**Pull request by aricart**
This is a similar error to the one we were discussing on #292. Effectively, this is another upvote.
The fix is c/p from #292. Before the end of the year we need to cut a 2.0, with the changes discussed.

### CODE REVIEW

File: test/basics.js [hunk 1]
```diff
@@ -877,4 +877,87 @@ describe('Basics', function() {
...
+
+            nc.flush(() => {
+                nc.requestOne(subj, 'some string', {queue: 'worker'}, 1000, h);
+                nc.requestOne(subj, {}, {queue: 'worker'}, 1000, h);
+                nc.requestOne(subj, 123, {queue: 'worker'}, 1000, h);
+            });
+        });
+    });
+
+    it('json requests', function(done) {
```
Comment (kozlovic) [inline test/basics.js]: Looks to me that this is the exact same test than the one just above.

File: test/shifter.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright 019 The NATS Authors
```
Comment (kozlovic) [inline test/shifter.js]: Date is wrong

File: test/shifter.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        const a = new Array(0);
+        const b = NATS.callbackShifter(0, a);
+        b.should.have.length(0);
+    });
+
+    it('should handle just fun', () => {
+        const a  = [function () {}];
+        const b = NATS.callbackShifter(10, a);
+        b.should.have.length(10);
+        for(let i=0; i < 8; i++) {
```
Comment (kozlovic) [inline test/shifter.js]: `<=8`

File: lib/nats.js [hunk 4]
```diff
@@ -1905,6 +1909,40 @@ Client.prototype.requestOne = function(subject, opt_msg, opt_options, timeout, c
     return this.request(subject, opt_msg, opt_options, callback);
 };
 
+// finds the last index that satisfies the testFun in array a
+function lastIndexOf(a, testFun) {
+    const idx = a.length-1;
+    for(let i=idx; i > -1; i--) {
```
Comment (kozlovic) [inline lib/nats.js]: `i >= 0`?

File: lib/nats.js [hunk 5]
```diff
@@ -1905,6 +1909,40 @@ Client.prototype.requestOne = function(subject, opt_msg, opt_options, timeout, c
...
+// returns an array that shifts the last function in the
+// array to the last element of the array
+function callbackShifter(maxArgs, args) {
+    // last argument is expected to be a function
+    const a = new Array(maxArgs);
+    const idx = lastIndexOf(args, (v) => typeof v === 'function');
+
+    // found a function, at the last index
+    if (idx !== -1) {
+        a[maxArgs - 1] = args[idx];
```
Comment (kozlovic) [inline lib/nats.js]: Looks to me that you would lose the last element if that is not a function, and you would still keep the function at its original position. Suppose you have:
`1, 2, 3, F, 4, 5` where `F` is a function. The `lastIndexOf()` would return idx 3. In array `a` you would put `a[5]='F'`, and then copy only from 0 to 2, so `a` would result in: `1, 2, 3, _,  _, F` if I understand your code.. (here the `_` means that there is no value copied from `args`, so undefined I guess.

Comment (aricart) [reply lib/nats.js]: the callbacks always end in a function, so arguments 4, 5 and not valid, and are dropped - this is is correct.

Comment (kozlovic) [reply lib/nats.js]: but maxArgs is 6 in the above case, no?

**Pull request closed without merging**